File: Problem_1.py
# Using Heap
# Time Complexity: O(n log(k))
# Space Complexity: O(n)
class FreqEle:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        
    def __lt__(self, other):
        if self.value == other.value:
            return self.key > other.key
        return self.value < other.value

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        if k == len(nums):
            return nums
        hashmap = {}
        for num in nums:
            hashmap[num] = hashmap.get(num, 0) + 1
        
        pq = []
        for key, value in hashmap.items():
            heapq.heappush(pq, FreqEle(key, value))
            if len(pq) > k:
                heapq.heappop(pq)
                
        result = []
        print(pq)
        while pq:
            result.append(heapq.heappop(pq).key)
        return result
                

# Using Buckets
# Time Complexity: O(n)
# Space Complexity: O(n)
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        if k == len(nums):
            return nums
        hashmap = {}
        for num in nums:
            hashmap[num] = hashmap.get(num, 0) + 1
        print(hashmap)
        
        buckets = [None] * (len(nums) + 1)
        for key, value in hashmap.items():
            if buckets[value] == None:
                buckets[value] = list()
            buckets[value].append(key)
       
        result = []
        for i in range(len(buckets)-1,-1,-1):
            if buckets[i] != None:
                for j in range(len(buckets[i])-1, -1, -1):
                    if k > 0:
                        result.append(buckets[i][j])
                        k -= 1
        return result
            
        

File: problem1.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<unordered_set>

using namespace std;

/*
// Time Complexity : O(nlogn)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
Build hashmap for frequency 
Insert them into heap on the basis of their frequency 
pop the first k elements
*/


struct CustomComparator{
    bool operator()(vector<int>& n1,vector<int>& n2){
        return n1.at(1)<n2.at(1); //decreasing order.
    }
};


class Solution {
    
public:
    vector<int> topKFrequent(vector<int>& nums, int k){
        unordered_map<int,int> hash_map{};
        priority_queue<vector<int>,vector<vector<int>>,CustomComparator> pq;
        int sz = nums.size();
        for(int& num:nums){
            if(hash_map.find(num) == hash_map.end()){
                hash_map[num] = 0;
            }
            hash_map[num]++;
        }
        for(auto x:hash_map){
            vector<int> temp{x.first,x.second};
            pq.push(temp);
        }
        vector<int> res{};
        for(int i{};i<k;++i){
            res.push_back(pq.top().at(0));
            pq.pop();
        }
        return res;
    }
};